### PR TITLE
Tpetra: Improving std::move-ability of BlockMultiVector

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #11931
Tests:Added
Stakeholder Feedback: See #11931

### CODE REVIEW

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp [hunk 1]
```diff
@@ -320,15 +320,25 @@ class BlockMultiVector :
...
   makePointMap (const map_type& meshMap, const LO blockSize);
 
+  /// \brief Create and return the point Map corresponding to the
+  ///   given mesh Map and block size.
+  ///
+  /// This is a class ("static") method so that you can make and reuse
+  /// a point Map for creating different BlockMultiVector instances,
+  /// using the more efficient four-argument constructor.
+  static Teuchos::RCP<const map_type>
+  makePointMapRCP (const map_type& meshMap, const LO blockSize);
```
Comment (cwpearson) [inline packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp:334]: Let's specify that this is an owning RCP

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp [hunk 2]
```diff
@@ -320,15 +320,25 @@ class BlockMultiVector :
...
+
   /// \brief Get this BlockMultiVector's (previously computed) point Map.
   ///
   /// It is always valid to call this method.  A BlockMultiVector
   /// always has a point Map.  We do not compute the point Map lazily.
-  map_type getPointMap () const {
-    return pointMap_;
+  const map_type getPointMap () const {
```
Comment (cwpearson) [inline packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp:340]: If there's a motivation for returning `const` here, let's explain in the function docs

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp [hunk 3]
```diff
@@ -87,8 +87,8 @@ BlockMultiVector (const map_type& meshMap,
                   const LO numVecs) :
   dist_object_type (Teuchos::rcp (new map_type (meshMap))), // shallow copy
   meshMap_ (meshMap),
-  pointMap_ (makePointMap (meshMap, blockSize)),
-  mv_ (Teuchos::rcpFromRef (pointMap_), numVecs), // nonowning RCP is OK, since pointMap_ won't go away
+  pointMap_ (makePointMapRCP (meshMap, blockSize)),
+  mv_ (pointMap_, numVecs), // nonowning RCP is OK, since pointMap_ won't go away
   blockSize_ (blockSize)
 {}
```
Comment (cwpearson) [inline packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp:93]: Can we assert any expectations about owning-ness of any of these RCPs here?

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp [hunk 4]
```diff
@@ -213,7 +212,7 @@ makePointMap (const map_type& meshMap, const LO blockSize)
 
   if (meshMap.isContiguous ()) {
     return map_type (gblNumPointMapInds, lclNumPointMapInds, indexBase,
-                     meshMap.getComm ());
+                                     meshMap.getComm ());
```
Comment (cwpearson) [inline packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp:215]: Formatting is a bit strange but I'll accept it

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp [hunk 5]
```diff
@@ -233,11 +232,57 @@ makePointMap (const map_type& meshMap, const LO blockSize)
       }
     }
     return map_type (gblNumPointMapInds, lclPointGblInds (), indexBase,
-                     meshMap.getComm ());
+                 meshMap.getComm ());
+
```
Comment (cwpearson) [inline packages/tpetra/core/src/Tpetra_BlockMultiVector_def.hpp:236]: Formatting is a bit strange but I'll accept it.

**Review by cwpearson — CHANGES_REQUESTED** [review]
In principle it looks good

File: packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp (hunk 2 above)
Comment (cwpearson) [reply packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp:340]: Then I think we want to reuturn `Teuchos::RCP<const map_type>`?

Comment (cwpearson) [reply packages/tpetra/core/src/Tpetra_BlockMultiVector_decl.hpp:340]: Ah I follow now

**Review by cwpearson — DISMISSED** [review]

**Comment by sebrowne** [comment]
Off the top of my head, this looks more like a bad interaction between Tpetra and KokkosKernels in the context of an installation.  @bartlettroscoe can you weigh in?  I'm not super well versed in the way that the installation files get generated, but my suspicion is that KokkosKernels is not installing files that it's supposed to/tpetra is depending on a target that it should not/the generated installation is wrong for a TriBITS reason.

**Comment by bartlettroscoe** [comment]
> Off the top of my head, this looks more like a bad interaction between Tpetra and KokkosKernels in the context of an installation. @bartlettroscoe can you weigh in?

The problem is described in #11931?  It might take me a bit to get up to speed on that but I can dig in if it is high priority.

**Comment by sebrowne** [comment]
@bartlettroscoe that's the issue that this PR is addressing, but I believe what @csiefer2 is saying is that the failures don't appear related.  I took a look and I'm inclined to agree, but I think that they are highlighting a valid failure inside of the installation (some sort of bad install logic in either Tpetra or KokkosKernels).

See the results under TEST_2 heading on this page: https://trilinos-cdash.sandia.gov/test/28205568

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe that's the issue that this PR is addressing, but I believe what @csiefer2 is saying is that the failures don't appear related. I took a look and I'm inclined to agree, but I think that they are highlighting a valid failure inside of the installation (some sort of bad install logic in either Tpetra or KokkosKernels).
> 
> See the results under TEST_2 heading on this page: https://trilinos-cdash.sandia.gov/test/28205568

@csiefer2 and @sebrowne, that error has nothing to do with "installation" per say.  That failure in the test `TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree` pulls in the `KokkosKernelsConfig.cmake` file from the build tree using the include dirs declared for the **build tree** (not the install tree) by KokkosKernels.  The error shown in that test is:

```
CMake Error in CMakeLists.txt:
  Imported target "Tpetra::all_libs" includes non-existent path

    "/scratch/trilinos/workspace/Trilinos_PR_cuda-11.4.2-uvm-off/pull_request_test/packages/kokkos-kernels/batched"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.
```

The problem is that the **source** directory  `packages/kokkos-kernels/batched` does not seem to exist for some reason for these PR builds.  But it seems that this source directory **does exist** in 'develop' in the Trilinos repo as shown by:

* https://github.com/trilinos/Trilinos/tree/develop/packages/kokkos-kernels/batched

Therefore, I can't explain this failure.

NOTE: The test `TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree`  passes in other PR builds recently as shown in this query:

* https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&begin=2023-06-07&end=2023-06-12&filtercount=2&showfilters=1&filtercombine=and&field1=groupname&compare1=61&value1=Pull%20Request&field2=testname&compare2=61&value2=TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree

Can we look on those machines and see if that directory:

* `/scratch/trilinos/workspace/<jenkins-base-dir>/pull_request_test/packages/kokkos-kernels/batched`

exists or not?  (I don't have the permissions to log onto those machines.)

I know the build directory is blown away after each PR build but the source directory should still be maintained, correct?

**Comment by bartlettroscoe** [comment]
> @bartlettroscoe that's the issue that this PR is addressing, but I believe what @csiefer2 is saying is that the failures don't appear related. I took a look and I'm inclined to agree

@sebrowne and @csiefer2, I agree, this PR and those failures are likely unrelated.  I don't see how the changes in this PR could be causing that test to fail.  We could pull this branch locally and fire off one of the PR builds locally and and run this test and see what happens?  That should be trivial to do with the tool:

* https://gitlab-ex.sandia.gov/rabartl/run_trilinos_pr_builds

You would run it with something like:

```
$ ./run_trilinos_pr_builds.py -p TrilinosInstallTests -f off -j 16 -t 4 clang-11.0.1
```

**Comment by sebrowne** [comment]
@bartlettroscoe those appear to be build directories, and the 'batched' directory is not present.  I think that's not being built, even though it probably should be?

As a side note, that directory is absolutely not present, I had to correct the path.  /scratch/trilinos/workspace does not exist (it's /scratch/trilinos/jenkins/...).  Maybe the path determination is the problem?

**Comment by bartlettroscoe** [comment]
Let me try to reproduce this failure offline and see what I can figure out.

**Comment by bartlettroscoe** [comment]
FYI: I was able to reproduce the problem locally (see below for details).  I will now see what is causing this and what needs to be done to fix this.  (And I will push the fix to this branch and then create a KokkosKernels PR against the main repo.)

<details>

<summary><b>Reproduction of problematic KokkosKernelsTargets.cmake file with bad kokkos-kernels/batched include dir</b> (click to expand)</summary>

Reproducing the failing test TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree in Trilinos PR #11966.

On the machine 'crf450', the Trilinos repos are:

```
$ ssh crf450

$ cd ~/Trilinos.base2/Trilinos/

$ gitdist-status 
------------------------------------------------------------------------------------------------
| ID | Repo Dir                         | Branch          | Tracking Branch        | C | M | ? |
|----|----------------------------------|-----------------|------------------------|---|---|---|
|  0 | Trilinos (Base)                  | csiefer-c9e0176 | github/csiefer-c9e0176 |   |   |   |
|  1 | packages/framework/GenConfig     | HEAD            |                        |   |   |   |
|  2 | packages/framework/son-ini-files | master          | glex-tril/master       |   |   |   |
|  3 | packages/framework/srn-ini-files | master          | cgl-tril/master        |   |   |   |
|  4 | run_trilinos_pr_builds           | master          | glex-rabartl/master    |   |   |   |
------------------------------------------------------------------------------------------------

$ gitdist-repo-versions 
*** Base Git Repo: Trilinos
51e55d800451f051d646b8caa0ba08b59e5f5f51 [Mon Jun 12 15:07:52 2023 -0600] <csiefer@sandia.gov>
Tpetra: Carl's fixes
*** Git Repo: packages/framework/GenConfig
924a08af66f0a0573b5dd1128179731489339aec [Wed Jun 15 15:55:48 2022 +0000] <eharvey@sandia.gov>
Merge branch 'del-gen-symlinks' into 'master'
*** Git Repo: packages/framework/son-ini-files
341d4592119e317ff218ce41c6cb029f3d68f5f6 [Tue May 30 19:05:59 2023 +0000] <sebrown@sandia.gov>
Merge branch 'sebrown/container' into 'master'
*** Git Repo: packages/framework/srn-ini-files
da821b1eb596b57a4f4672b94323e41e821f4997 [Thu Jun 8 14:31:49 2023 -0600] <sebrown@sandia.gov>
Merge branch 'sebrown/cts1' into 'master'
*** Git Repo: run_trilinos_pr_builds
a689be32cdf668e4ce6bfa5e8be91b651846968b [Tue Mar 28 06:47:39 2023 -0600] <rabartl@sandia.gov>
Update set of PR builds as of 2023-03-28 (trilinos/Trilinos#11730)
```

Now, genetating the build directory and doing the configure:

```
$ ssh crf450

$ cd ~/Trilinos.base2/BUILDS/PR/pr_builds/

$ ./run_trilinos_pr_builds.py -p TrilinosInstallTests -f off -j 16 -t 4 -g clang-11.0.1

$ cd clang-11.0.1/

$ rm -r packages cmake_packages

$ . load-env.sh

$ time ./do-configure &> configure.out 

real    0m14.898s
user    0m12.457s
sys     0m9.812s
```

That generated:

```
$ ls packages/kokkos-kernels/
blas  CMakeFiles  cmake_install.cmake  CTestTestfile.cmake  graph  KokkosKernels_config.h  Makefile  sparse

$ ~/tree.py -cf cmake_packages/KokkosKernels/
  KokkosKernels/
  +-KokkosKernelsConfig.cmake
  +-KokkosKernelsTargets.cmake

$ grep batched cmake_packages/KokkosKernels/KokkosKernelsTargets.cmake 
  INTERFACE_INCLUDE_DIRECTORIES "/ascldap/users/rabartl/Trilinos.base2/BUILDS/PR/pr_builds/clang-11.0.1;/ascldap/users/rabartl/Trilinos.base2/BUILDS/PR/pr_builds/clang-11.0.1/packages/kokkos-kernels/batched;..."
```

So the above shows that we do indeed get the directory:

```
  <build-dir>/packages/kokkos-kernels/batched
```

added to the `INTERFACE_INCLUDE_DIRECTORIES` property.  How does this happen?

</details>

**Comment by bartlettroscoe** [comment]
@sebrowne and @csiefer2, the failing test was caused by a defect in the KokkosKernels CMake build system and is fixed in the commit https://github.com/bartlettroscoe/kokkos-kernels/commit/aee776e44588481b2b99cb047c1b572f77c66ef7 and the matching KokkosKernels PR:

* https://github.com/kokkos/kokkos-kernels/pull/1867

I verified that this commit fixed the local running of the test `TrilinosInstallTests_simpleBuildAgainstTrilinos_by_package_build_tree` and I have high confidence that this will pass in the updated PR build that will run next.

File: packages/kokkos-kernels/batched/CMakeLists.txt [hunk 6]
```diff
@@ -16,6 +16,15 @@ IF (NOT KokkosKernels_ENABLE_COMPONENT_BLAS)
   LIST(APPEND SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/batched/KokkosBatched_Util.cpp)
 ENDIF()
 
-# Adding unit-tests
-KOKKOSKERNELS_INCLUDE_DIRECTORIES(${CMAKE_CURRENT_BINARY_DIR}/batched)
-KOKKOSKERNELS_INCLUDE_DIRECTORIES(REQUIRED_DURING_INSTALLATION_TESTING ${CMAKE_CURRENT_SOURCE_DIR}/batched)
+IF(KokkosKernels_ENABLE_TESTS OR KokkosKernels_ENABLE_TESTS_AND_PERFSUITE)
```
Comment (ndellingwood) [inline packages/kokkos-kernels/batched/CMakeLists.txt:19]: Approving kokkos-kernels fix, thanks @bartlettroscoe !

**Review by ndellingwood — APPROVED** [review]

**Pull request merged**